src/Containers/container.jl

```diff
@@ -113,13 +113,51 @@ function container(f::Function, indices, ::Type{SparseAxisArray})
     mappings = Base.Generator(I -> I => f(I...), indices)
     # Same as `Dict(mapping)` but it will error if two indices are the same.
     data = NoDuplicateDict(mappings)
-    dict = data.dict
-    if isempty(dict)
-        # If `dict` is empty, it was not able to determine the type
-        # of the key hence the type of `dict` is `Dict{Any, Any}`.
-        # This is an issue since `SparseAxisArray` needs the key
-        # type to be a tuple.
-        dict = Dict{_eltype_or_any(indices),Any}()
-    end
-    return SparseAxisArray(dict)
+    return _sparseaxisarray(data.dict, f, indices)
+end
+
+# The NoDuplicateDict was able to infer the element type.
+_sparseaxisarray(dict::Dict, ::Any, ::Any) = SparseAxisArray(dict)
+
+function _container_dict(
+    K::Type{<:NTuple{N,Any}},
+    f::Function,
+    ::Type{<:NTuple{N,Any}},
+) where {N}
+    ret = Base.return_types(f, K)
+    V = length(ret) == 1 ? first(ret) : Any
+    return Dict{K,V}()
+end
+
+function _container_dict(
+    ::Union{Any,Type{Union{}}},
+    ::Function,
+    K::Type{<:NTuple{N,Any}},
+) where {N}
+    return Dict{K,Any}()
+end
+
+function _container_dict(::Any, ::Any, K::Type{<:NTuple{N,Any}}) where {N}
+    return Dict{K,Any}()
+end
+
+function _container_dict(
+    ::Type{Union{}},
+    ::Function,
+    K::Type{<:NTuple{N,Any}},
+) where {N}
+    return Dict{K,Any}()
+end
+
+_default_eltype(x::NestedIterator) = Base.@default_eltype nested(x.iterators...)
```

Comment:
Why do we need this ? Isn't the condition ignored by `Base.default_eltype` ?

Reply:
No, because the condition is part of the `iterate` interface.

Reply:
Yes but the compiler can't tell whether the condition is true or false so it assumes it might be true and add the returned type to the `Union`. At leas that's what I understand

Reply:
With the condition, it's more likely to return `Union{}` instead of inferring properly. They don't return the same thing.